crypto/objects/obj_xref.c

```diff
@@ -36,7 +36,12 @@ static int sigx_cmp(const nid_triple *const *a, const nid_triple *const *b)
     int ret;
 
     ret = (*a)->hash_id - (*b)->hash_id;
-    if (ret != 0)
+    /* consider a good algorithm match either one with identical digests
+     * or one with a signature algorithm indicating use of any, incl. no
+     * digest for proper operation (i.e., setting NID_undef for hash alg).
+     * SIG NID (pkey_id) must match in any case.
+     */
+    if ((ret != 0) && ((*b)->hash_id != NID_undef))
```

Comment:
It is slightly odd that the "b" side of the comparison is treated differently to the "a" side. So `sigx_cmp(a, b)` is not equivalent to `sigx_cmp(b, a)`

Reply:
That is because the "b" side carries the algorithms in the list of (sig,dgst,pkey) triplets "sigx_app" (the already registered algorithms) and the "a" side the triplet searched for: Only a `NID_undef` on that side means that a sig algorithm has already been registered not needing a digest.

Reply:
Please see new comment's explanation.